peps/pep-0724.rst

```diff
@@ -0,0 +1,305 @@
+PEP: 724
+Title: Stricter Type Guards
+Author: Rich Chiodo <rchiodo at microsoft.com>,
+        Eric Traut <erictr at microsoft.com>,
+        Erik De Bonte <erikd at microsoft.com>,
+Sponsor: Jelle Zijlstra <jelle.zijlstra@gmail.com>
+Discussions-To: https://mail.python.org/archives/list/typing-sig@python.org/thread/EMUD2D424OI53DCWQ4H5L6SJD2IXBHUL/
+Status: Draft
+Type: Standards Track
+Topic: Typing
+Content-Type: text/x-rst
+Created: 28-Jul-2023
+Python-Version: 3.13
+Post-History:
+
+
+Abstract
+========
+
+:pep:`647` introduced the concept of ``TypeGuard`` functions which return
+``True`` if their input parameter matches their target type. For example, a
+function that returns ``TypeGuard[str]`` is assumed to return ``True`` if and
+only if its input parameter is a ``str``. This allows type checkers to narrow
+types when a ``TypeGuard`` function returns ``True``.
+
+This PEP further refines :pep:`647` by allowing type checkers to also narrow
+types when a ``TypeGuard`` function returns ``False``.
+
+
+Motivation
+==========
+
+``TypeGuard``\s are used throughout Python libraries to allow a type checker
+to narrow the type of something when the ``TypeGuard`` returns ``True``.
+
+However, in the ``False`` case, :pep:`647` doesn't prescribe what the type
+might be:
+
+.. code-block:: python
+
+    def is_str(val: str | int) -> TypeGuard[str]:
+        return isinstance(val, str)
+
+    def func(val: str | int):
+        if is_str(val):
+            # Type checkers can assume val is a 'str' in this branch
+        else:
+            # Type here is not narrowed. It is still 'str | int'
+
+
+Specification
+=============
+
+Given a ``TypeGuard`` and a calling function like so:
+
+.. code-block:: python
+
+    def guard(x: P) -> TypeGuard[R]: ...
+
+    def func1(val: A):
+        if guard(val):
+            reveal_type(val)  # NP
+        else:
+            reveal_type(val)  # NN
+
+Each use of a ``TypeGuard`` involves five types:
+
+* P = ``TypeGuard`` parameter type
+* R = ``TypeGuard`` return type
+* A = Argument type (pre-narrowed)
+* NP = Narrowed type (positive)
+* NN = Narrowed type (negative)
```

Comment:
This introduction to the Specification seems a bit odd / doesn't flow very well -- you start with "Given a function" and then in the next paragraph define a series of types with no reference or connection to the previous paragraph.

Reply:
I'm sorry but I'm not sure which parts aren't referencing each other. Did you mean the Kangaroo | Koala example? Or the reference to the different types : 'P, R, A, NP, NN'

Reply:
Did the update clarify this at all?